pkg/instancegroups/rollingupdate.go

```diff
@@ -17,177 +17,319 @@ limitations under the License.
+					continue
+				}
+				// @step: wait for a slot to operate
+				<-bucket
 
-				g, err := NewRollingUpdateInstanceGroup(c.Cloud, group)
-				if err == nil {
-					err = g.RollingUpdate(c, instanceGroups, true, c.BastionInterval, c.ValidationTimeout)
+				// @step: determine the update strategy for this instancegroup
+				if err := c.DetermineGroupStratergy(x); err != nil {
+					return c.Errorf("unable to determine rollout strategy on instancegroup: %s, error: %v", name, err)
 				}
 
-				resultsMutex.Lock()
-				results[k] = err
-				resultsMutex.Unlock()
-			}(k, bastionGroup)
-		}
+				// @check we have instances inside the group which require updating
+				if len(x.NeedUpdate) == 0 {
+					c.Infof("skipping instancegroup: %s as no members have pending updates", name)
+					continue
+				}
 
-		wg.Wait()
-	}
+				worker.Add(1)
+				go func(group *cloudinstances.CloudInstanceGroup) {
+					defer func() {
+						worker.Done()
+						bucket <- Signal
+					}()
+					update := &RollingUpdateInstanceGroup{Update: c, CloudGroup: group}
+					// @step: perform a rollout on the group
+					if err := update.RollingUpdate(rctx, list); err != nil {
+						// return the error for the next iteration to pick up; so the channel is
+						// buffered so non-blocking here
+						errorCh <- err
+					}
+				}(x)
+			}
 
-	// Upgrade master next
-	{
-		var wg sync.WaitGroup
+			return nil
+		}(); err != nil {
+			errorCh <- err
+			return
+		}
+		// @step: wait for all the routines to finish
+		worker.Wait()
+		doneCh <- Signal
+	}()
 
-		// We run master nodes in series, even if they are in separate instance groups
-		// typically they will be in separate instance groups, so we can force the zones,
-		// and we don't want to roll all the masters at the same time.  See issue #284
-		wg.Add(1)
+	return resultCh
+}
 
-		go func() {
-			for k := range masterGroups {
-				resultsMutex.Lock()
-				results[k] = fmt.Errorf("function panic masters")
-				resultsMutex.Unlock()
-			}
+// UpdateCluster is responsible for updating the kops cluster
+func (c *RollingUpdateCluster) UpdateCluster(ctx context.Context) error {
+	// @NOTE i'm not sure there is a way of cancelling this operation??
+	applyCmd := &cloudup.ApplyClusterCmd{
+		Clientset:       c.Clientset,
+		Cluster:         c.Cluster,
+		DryRun:          false,
+		InstanceGroups:  nil,
+		MaxTaskDuration: cloudup.DefaultMaxTaskDuration,
+		Models:          nil,
+		OutDir:          ".",
+		TargetName:      cloudup.TargetDirect,
+	}
 
-			defer wg.Done()
+	return applyCmd.Run()
+}
 
-			for k, group := range masterGroups {
-				g, err := NewRollingUpdateInstanceGroup(c.Cloud, group)
-				if err == nil {
-					err = g.RollingUpdate(c, instanceGroups, false, c.MasterInterval, c.ValidationTimeout)
-				}
+// Infof is used to provide details about the on-going rollout
+func (c *RollingUpdateCluster) Infof(message string, opts ...interface{}) {
+	glog.Infof(fmt.Sprintf(message, opts...))
```

Comment:
Do we need this one?  It feels unnecessary, and it adds overhead because we're always formatting the message even if it won't be printed